File: javatest/omnia/data/cache/WeakCacheTest.java
package omnia.data.cache;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth8.assertThat;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class WeakCacheTest {
  private final WeakCache<String, String> testSubject = new WeakCache<>();

  @Test
  public void contains_whenNotContained_returnsFalse() {
    assertThat(testSubject.contains("unknown")).isFalse();
  }

  @Test
  public void contains_whenContained_returnsTrue() {
    String value = "value";

    testSubject.getOrCache("key", () -> value);

    assertThat(testSubject.contains("key")).isTrue();
  }

  @Test
  public void getOrCache_whenNotContained_returnsSuppliedInstance() {
    String value = "value";

    assertThat(testSubject.getOrCache("key", () -> value)).isSameInstanceAs(value);
  }

  @Test
  public void getOrCache_whenContained_returnsCachedInstance() {
    @SuppressWarnings("RedundantStringConstructorCall") // we're testing reference behaviors
    String cachedValue = new String("cached");
    @SuppressWarnings("RedundantStringConstructorCall") // we're testing reference behaviors
    String suppliedValue = new String("supplied");

    testSubject.getOrCache("key", () -> cachedValue);

    assertThat(testSubject.getOrCache("key", () -> suppliedValue)).isSameInstanceAs(cachedValue);
  }

  @Test
  public void get_whenNotCached_isNotPresent() {
    assertThat(testSubject.get("key")).isEmpty();
  }

  @Test
  public void get_whenCached_isPresent() {
    String value = "value";

    testSubject.getOrCache("key", () -> value);

    assertThat(testSubject.get("key")).isPresent();
    assertThat(testSubject.get("key")).hasValue(value);
  }
}

File: tests/omnia/data/cache/WeakCacheTest.kt
package omnia.data.cache

import com.google.common.truth.Truth.assertThat
import kotlin.test.Test

class WeakCacheTest {

  private val testSubject = WeakCache<String, String>()

  @Test
  fun contains_whenNotContained_returnsFalse() {
    assertThat(testSubject.contains("unknown")).isFalse()
  }

  @Test
  fun contains_whenContained_returnsTrue() {
    val value = "value"
    testSubject.getOrCache("key") { value }
    assertThat(testSubject.contains("key")).isTrue()
  }

  @Test
  fun getOrCache_whenNotContained_returnsSuppliedInstance() {
    val value = "value"
    assertThat(testSubject.getOrCache("key") { value }).isSameInstanceAs(value)
  }

  @Test
  fun getOrCache_whenContained_returnsCachedInstance() {
    val cachedValue = "cached"
    val suppliedValue = "supplied"
    testSubject.getOrCache("key") { cachedValue }
    assertThat(testSubject.getOrCache("key") { suppliedValue }).isSameInstanceAs(cachedValue)
  }

  @Test
  fun get_whenNotCached_isNotPresent() {
    assertThat(testSubject["key"]).isNull()
  }

  @Test
  fun get_whenCached_isPresent() {
    val value = "value"
    testSubject.getOrCache("key") { value }
    assertThat(testSubject["key"]).isEqualTo(value)
  }
}